framework/src/wifi_manager/wifi_manager.c

```diff
@@ -489,284 +595,384 @@ wifi_manager_result_e wifi_manager_init(wifi_manager_cb_s *wmcb)
+		WIFIMNG_DEC_NUM_CLIENT;
+		_handle_user_cb(CB_EVT_LEAVE, NULL);
+	} else if (msg->event == EVT_DEINIT) {
+		res = _wifimng_stop_softap();
+		if (res != WIFI_MANAGER_SUCCESS) {
+			// CRITICAL ERROR
+		}
+		res = _wifimng_deinit();
+		if (res != WIFI_MANAGER_SUCCESS) {
+			// CRITICAL ERROR
+		}
+		WIFIMNG_SET_STATE(WIFIMNG_INIT);
+	} else {
+		return WIFI_MANAGER_FAIL;
 	}
-	wifi_mutex_release(w_mutex);
 
-	wifi_mutex_acquire(w_info_mutex, WIFI_UTILS_FOREVER);
-	ndbg("w_mutex->semcount: %d, w_mutex->flags: %d\n", w_mutex->semcount, w_mutex->flags);
+	return WIFI_MANAGER_SUCCESS;
+}
 
-	result = wifi_mutex_destroy(w_mutex);
-	if (result != WIFI_UTILS_SUCCESS) {
-		ndbg("wifi_mutex_destroy fail %d", result);
-		ret = WIFI_MANAGER_FAIL;
-	}
 
-	if (ret != WIFI_MANAGER_FAIL) {
-		g_manager_info.mode = WIFI_NONE;
-	} else {
-		g_manager_info.mode = WIFI_FAILURE;
+wifi_manager_result_e _handle_scanning(_wifimng_msg_s *msg)
+{
+	WM_LOG_HANDLER_START;
+	if (msg->event == EVT_SCAN_DONE) {
+		_handle_user_cb(CB_EVT_SCAN, msg->param);
+		WIFIMNG_RESTORE_STATE;
+		return WIFI_MANAGER_SUCCESS;
 	}
+	return WIFI_MANAGER_FAIL;
+}
 
-	if (w_mutex != NULL) {
-		free(w_mutex);
-		w_mutex = NULL;
-	}
-	wifi_mutex_release(w_info_mutex);
 
-	result = wifi_mutex_destroy(w_info_mutex);
-	if (result != WIFI_UTILS_SUCCESS) {
-		ndbg("w_info_mutex_destroy fail %d", result);
+void _handle_user_cb(_wifimng_usr_cb_type_e evt, void *arg)
+{
+	WM_LOG_START;
+	wifi_manager_cb_s *cbk = &g_manager_info.cb;
+	switch (evt) {
+	case CB_EVT_STA_CONNECTED:
+		printf("call sta connected event\n");
+		cbk->sta_connected();
+		break;
+	case CB_EVT_STA_CONNECT_FAIL:
+		printf("call sta connect fail event\n");
+		cbk->sta_connected();
+		break;
+	case CB_EVT_STA_DISCONNECT:
+		printf("call sta disconnected event\n");
+		cbk->sta_disconnected();
+		break;
+	case CB_EVT_JOIN:
+		printf("call sta join event\n");
+		cbk->softap_sta_joined();
+		break;
+	case CB_EVT_LEAVE:
+		printf("call sta leave event\n");
+		cbk->softap_sta_left();
+		break;
+	case CB_EVT_SCAN:
+		printf("call sta scan event\n");
+		// convert scan data.
+		wifi_manager_scan_info_s *info = NULL;
+		wifi_utils_scan_list_s *list = (wifi_utils_scan_list_s *)arg;
+		if (list) {
+			wifi_manager_result_e res = _convert_scan_info(&info, (wifi_utils_scan_list_s *)arg);
+			if (res != WIFI_MANAGER_SUCCESS) {
+				printf("parsing error\n");
+			}
+			cbk->scan_ap_done(&info, WIFI_SCAN_SUCCESS);
+			_free_scan_info(info);
+		} else {
+			cbk->scan_ap_done(NULL, WIFI_SCAN_FAIL);
+		}
+		break;
+	default:
+		ndbg("Invalid State\n");
+		break;
 	}
-	if (w_info_mutex != NULL) {
-		free(w_info_mutex);
-		w_info_mutex = NULL;
+}
+
+
+wifi_manager_result_e _handle_request(_wifimng_msg_s *msg)
+{
+	WM_LOG_HANDLER_START;
+	wifi_manager_result_e res = WIFI_MANAGER_FAIL;
+
+	LOCK_WIFIMNG;
+	res = g_handler[WIFIMNG_GET_STATE](msg);
+	if (res != WIFI_MANAGER_SUCCESS) {
+		ndbg("Wi-Fi manager fail: state error\n");
```

Comment:
Developers may want more specific error messages.

Reply:
About error code, it will be compensated at the next pull reqeust.